src/main/scala/intellij/haskell/module/HaskellModuleBuilder.scala

```diff
@@ -218,12 +223,18 @@ object HaskellModuleBuilder {
     }
   }
 
-  private def downloadHaskellPackageSources(project: Project, stackPath: String, haskellPackages: Seq[HaskellPackageInfo], progressIndicator: ProgressIndicator) = {
+  private def downloadHaskellPackageSources(project: Project, libRoot: File, stackPath: String, haskellPackages: Seq[HaskellPackageInfo], progressIndicator: ProgressIndicator) = {
     val step = 0.8 / haskellPackages.size
     var progressFraction = InitialProgressStep
-    haskellPackages.flatMap { packageInfo =>
-      val fullName = packageInfo.name + "-" + packageInfo.version
-      val stdErr = CommandLine.runProgram(Some(project), project.getBasePath + File.separator + LibName, stackPath, Seq("unpack", fullName), 10000, captureOutputToLog = true, logErrorAsInfo = true).map(_.getStderr)
+
+    libRoot.listFiles().filterNot(subDir => haskellPackages.map(_.dirName).contains(subDir.getName)).foreach(FileUtil.delete)
+
+    haskellPackages.filter(packageInfo => {
```

Comment:
Hi @rikvdkleij, I have to say the current logic in my code is right, I can not see any part in the console complain about re-downloading packages which already downloaded. And the screenshot below can give the evidence that all the packages are downloaded in Fab 25th, and they will never be downloaded again only if I remove them.

<img width="424" alt="screen shot 2017-02-26 at 2 41 27 pm" src="https://cloud.githubusercontent.com/assets/6234553/23337721/b4bf46e8-fc31-11e6-8768-369b31aee9c3.png">
>Also, if you test it, you can look into the console to see that existing packages are downloaded again.